### PR TITLE
Fix guards for field access through properties

### LINKED ISSUE #118786: 43 Dynamo test are failing with "'bool' object has no attribute '__closure__'".
43 Dynamo test are failing with "'bool' object has no attribute '__closure__'".

## Repro

`PYTORCH_TEST_WITH_DYNAMO=1 pytest test_quantization.py -v -k test_avg_pool2d_nhwc`

You will need to remove the skip or expectedFailure before running the repro command.
This may be just removing a line in
[dynamo_test_failures.py](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/dynamo_test_failures.py)

## Failing tests

Here's a comprehensive list of tests that fail (as of this issue) with the above message:
<details>
<summary>Click me</summary>
```
test_quantization.py::TestQuantizedOps.test_avg_pool2d_nhwc
test_quantization.py::TestQuantizedOps.test_cat_nhwc
test_quantization.py::TestQuantizedOps.test_channel_shuffle
test_quantization.py::TestQuantizedOps.test_equal
test_quantization.py::TestQuantizedOps.test_interpolate
test_quantization.py::TestQuantizedOps.test_leaky_relu_observed_output
test_quantization.py::TestQuantizedOps.test_mul_scalar_relu
test_quantization.py::TestQuantizedOps.test_sigmoid
test_quantization.py::TestQuantizedOps.test_sigmoid_non_observed
test_quantization.py::TestQNNPackOps.test_qnnpack_maxpool2d
test_quantization.py::TestQuantizedLinear.test_qlinear_unpack
test_quantization.py::TestQuantizedConv.test_qconv3d_unpack
test_quantization.py::TestComparatorOps.test_compare_tensor_tensor
test_quantization.py::TestFakeQuantizeOps.test_fake_quant_per_channel_qparam_range
test_quantization.py::TestFakeQuantizeOps.test_fq_module_per_tensor
test_quantization.py::TestFakeQuantizeOps.test_learnable_backward_per_tensor_cpu
test_quantization.py::TestQuantizedOps.test_add_scalar_relu
test_quantization.py::TestQuantizedOps.test_avg_pool2d
test_quantization.py::TestQuantizedOps.test_avg_pool3d
test_quantization.py::TestQuantizedOps.test_avg_pool3d_nhwc
test_quantization.py::TestQuantizedOps.test_cat
test_quantization.py::TestQuantizedOps.test_hardtanh
test_quantization.py::TestQuantizedOps.test_interpolate3d
test_quantization.py::TestQuantizedOps.test_max_pool1d
test_quantization.py::TestQuantizedOps.test_max_pool2d
test_quantization.py::TestQuantizedOps.test_max_pool2d_nhwc
test_quantization.py::TestQuantizedOps.test_qcelu
test_quantization.py::TestQuantizedOps.test_qclamp
test_quantization.py::TestQuantizedOps.test_qelu
test_quantization.py::TestQuantizedOps.test_qthreshold
test_quantization.py::TestQNNPackOps.test_qnnpack_add
test_quantization.py::TestQNNPackOps.test_qnnpack_mul
test_quantization.py::TestQNNPackOps.test_qnnpack_relu
test_quantization.py::TestQuantizedLinear.test_qlinear_qnnpack_free_memory_and_unpack
test_quantization.py::TestQuantizedConv.test_qconv1d_unpack
test_quantization.py::TestQuantizedConv.test_qconv2d_unpack
test_quantization.py::TestComparatorOps.test_compare_tensor_scalar
test_quantization.py::TestFakeQuantizeOps.test_fixed_qparams_fq_module
test_quantization.py::TestFakeQuantizeOps.test_forward_per_channel
test_quantization.py::TestFakeQuantizeOps.test_forward_per_tensor
test_quantization.py::TestFakeQuantizeOps.test_learnable_forward_per_channel_cpu
test_quantization.py::TestFakeQuantize.test_fq_module_per_channel
test_nn.py::TestFusionEval.test_fuse_module_eval_numerics
```
</details>

cc @ezyang @msaroufim @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by aorenste**
When building guards which went through a property we were analyzing the property using getattr_static but the guard wasn't built using getattr_static so if the property was "unusual" it generated misbehaved code which referenced a non-existent `__closure__` field.

Fixes #118786

Note that after this change some of the referenced tests are still failing with a different error - but getting further.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119719

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -727,6 +727,10 @@ class NO_SUCH_SUBOBJ:
                 unimplemented("UserDefined with non-function __getattr__")
 
         if isinstance(subobj, property):
+            # Rewrite the source being explicit about reading it statically.
+            if self.source:
+                source = AttrSource(self.source, name, get_static=True)
+                source = AttrSource(source, "fget")
```
Comment (oulgen) [inline torch/_dynamo/variables/user_defined.py:733]: curious, what's this fget here?

**Review by oulgen — APPROVED** [review]

File: torch/_dynamo/variables/user_defined.py (hunk 1 above)
Comment (aorenste) [reply torch/_dynamo/variables/user_defined.py:733]: Python property classes are made of three parts:
fget - the getter function
fset - the setter function
fdel - the `del` function

So if we're reading the value on the property we want to access the underlying 'fget' function. This is what the code below here was already doing (`subobj.fget`) but we weren't registering that fact with the guard.

File: test/dynamo/test_sources.py [hunk 2]
```diff
@@ -22,6 +22,29 @@ def test_is_local(self):
...
+
+        def func(elements):
+            if not elements.myprop:
+                return torch.tensor([1, 2, 3])
+            else:
+                return torch.tensor([4, 5, 6])
+
+        e = Elements()
+        a = func(e)
+        b = torch.compile(func)(e)
```
Comment (zou3519) [inline test/dynamo/test_sources.py:45]: test/dynamo tests either use torch.compile(backend=eager) or torch._dynamo.optimize. using the full torch.compile is slow.

Comment (zou3519) [reply test/dynamo/test_sources.py:45]: Also, can we put fullgraph=True on this?

File: torch/_dynamo/source.py [hunk 3]
```diff
@@ -161,7 +162,9 @@ def guard_source(self):
         return self.base.guard_source()
 
     def name(self):
-        if not self.member.isidentifier():
+        if self.get_static:
+            return f"inspect.getattr_static({self.base.name()}, {self.member!r})"
```
Comment (zou3519) [inline torch/_dynamo/source.py:166]: What is the problem? that getattr_static returns something different from the other branch?

Comment (aorenste) [reply torch/_dynamo/source.py:166]: Correct. getattr_static() is a very different operation from getattr() so we need to make sure that when we're inspecting data using getattr_static() we express the guard the same way.

Comment (zou3519) [reply torch/_dynamo/source.py:166]: is getattr_static actually giving us the correct thing in this case?

Comment (aorenste) [reply torch/_dynamo/source.py:166]: I believe that (at least in the case I was looking at) getattr_static was correct. The code is checking for a property and if so then it's inlining the getter for that property. The problem was that if the inlined function built a guard based on the closure's captured data the guard was referring to the getattr()'d version of the property and then trying to refer to __closure__ off that.

File: test/dynamo/test_sources.py (hunk 2 above)
Comment (aorenste) [reply test/dynamo/test_sources.py:45]: https://github.com/pytorch/pytorch/pull/119921

**Pull request closed without merging**